src/wallet/rpcwallet.cpp

```diff
@@ -2534,43 +2534,64 @@ UniValue z_listunspent(const UniValue& params, bool fHelp)
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, expected string");
             }
             string address = o.get_str();
+            if (setAddress.count(address)) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER,
+                    string("Invalid parameter, duplicated address: ") + address);
+            }
+            setAddress.insert(address);
+
             auto zaddr = DecodePaymentAddress(address);
-            if (IsValidPaymentAddress(zaddr)) {
-                auto hasSpendingKey = boost::apply_visitor(HaveSpendingKeyForPaymentAddress(pwalletMain), zaddr);
-                if (hasSpendingKey) {
-                    zaddrs.insert(zaddr);
+            if (!IsValidPaymentAddress(zaddr)) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER,
+                    string("Invalid parameter, address is not a valid zaddr: ") + address);
+            }
+            auto sproutAddr = boost::get<libzcash::SproutPaymentAddress>(&zaddr);
+            if (sproutAddr != nullptr) {
+                if (!fIncludeWatchonly && !pwalletMain->HaveSproutSpendingKey(*sproutAddr)) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER,
+                        string("Invalid parameter, spending key for Sprout address does not belong to wallet: ") + address);
                 }
-            } else {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, string("Invalid parameter, address is not a valid zaddr: ") + address);
+                zaddrs.insert(*sproutAddr);
+                continue;
```

Comment:
I like doing `continue` rather than `else` if possible. This way, when (and if) Sprout is deprecated (and all support for it removed), lines 2548-2556 can be deleted, without touching the Sapling code path. 

In general I like the idea that when execution reaches a certain point, the set of possible states satisfies `A or B` and then we check if `B`, and if so, eliminate it (in this case by doing a `continue`), and then the set of states is _restricted_ to just `A` (`B` is no longer possible).